services/QuillLMS/client/app/bundles/Staff/components/activityForm/topics.tsx

```diff
@@ -39,12 +39,34 @@ const Topics = ({ activity, createNewTopic, topicOptions, handleTopicsChange, })
   }
 
   function onChangeTopics(topicId) {
-    let newTopicIds = [...activity.topic_ids, topicId]
+    let newTopicIds = [...activity.topic_ids]
     const topic = topicOptions.find(t => t.id === topicId)
     const existingOption = getSelectedOptionForLevel(topic.level)
+
     if (existingOption) {
       newTopicIds = newTopicIds.filter(id => id !== existingOption.id)
+      const existingParent = topicOptions.find(t => t.id === existingOption.parent_id)
+      if (existingParent) {
+        newTopicIds = newTopicIds.filter(id => id !== existingParent.id)
+        const existingGrandparent = topicOptions.find(t => t.id === existingParent.parent_id)
+        if (existingGrandparent) {
+          newTopicIds = newTopicIds.filter(id => id !== existingGrandparent.id)
+        }
+      }
+    }
+
+    if (!existingOption || existingOption.id !== topic.id) {
```

Comment:
The cyclomatic complexity of this function is rather large.

Reply:
I made some changes to try to simply it a little -- let me know what you think!

Reply:
This is looking much better.  One final refactor, consider:

```
if (x) {
  y
} else {
  y
  z
}
```

Since both branches contain `y` we can pull it out:
```
y

if (x) {
} else {
  z
}
```
Since that first conditional is empty now, we can do the negation:

```
y

if not (x) { z }
```
Which in this example reduces to:

```
 newTopicIds = removeExistingTopicAndParents(newTopicIds, previousTopic)

 if (previousTopic?.id !== newTopic.id) { newTopicIds = addNewTopicAndParents(newTopicIds, newTopic) }
```

Reply:
Gotcha. Thank you Brendan!